provisioning_client/adapters/hsm_client_riot.c

```diff
@@ -155,33 +136,39 @@ static const HSM_CLIENT_X509_INTERFACE x509_interface =
     hsm_client_riot_get_common_name
 };
 
-static int generate_root_ca_info(HSM_CLIENT_X509_INFO* riot_info, RIOT_ECC_SIGNATURE* tbs_sig)
+// Free the mbedtls_mpi members of the signature
+static void ecc_signature_destroy(RIOT_ECC_SIGNATURE* tbs_sig);
+
+static int generate_root_ca_info(HSM_CLIENT_X509_INFO* riot_info)
 {
     int result;
     uint8_t der_buffer[DER_MAX_TBS] = { 0 };
     DERBuilderContext der_ctx = { 0 };
     DERBuilderContext der_pri_ctx = { 0 };
     RIOT_STATUS status;
-
-    memcpy(&riot_info->ca_root_pub, eccRootPubBytes, sizeof(ecc_publickey));
-    memcpy(&riot_info->ca_root_priv, eccRootPrivBytes, sizeof(ecc_privatekey));
+    RIOT_ECC_SIGNATURE tbs_sig = { 0 };
 
     // Generating "root"-signed DeviceID certificate
     DERInitContext(&der_ctx, der_buffer, DER_MAX_TBS);
     DERInitContext(&der_pri_ctx, der_buffer, DER_MAX_TBS);
-
-    if (X509GetDeviceCertTBS(&der_ctx, &X509_ROOT_TBS_DATA, &riot_info->ca_root_pub) != 0)
+    if ((status = RiotCrypt_DeriveEccKey(&riot_info->ca_root_pub, &riot_info->ca_root_priv,
```

Comment:
I'm not sure I understand this part: where does the certificate come from? We need it to have it uploaded to the Provisioning service.

Reply:
Ln: 379 process_riot_key_info is what populates the actual x509 cert, this one of the helper functions to handle the ca-root part

Reply:
The DeviceID key/cert is held by the device (it may be persisted or derived on boot).  As per previous comment, the "root cert" is the cert corresponding to a key held by the root CA (not any device).  In the simplest case you have something like:

... > Root Cert (enrolled in DPS) > DeviceID Cert (on device) > ... > Alias Cert (leaf cert on device)

Reply:
see above comment